elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/metric/BaseMetric.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2019, Yahoo Inc.
+ * Licensed under the Apache License, Version 2.0
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.datastores.aggregation.metric;
+
+import com.yahoo.elide.core.EntityDictionary;
+import com.yahoo.elide.datastores.aggregation.annotation.Meta;
+import com.yahoo.elide.datastores.aggregation.annotation.MetricAggregation;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.lang.reflect.Constructor;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+/**
+ * {@link Metric} annotated by {@link MetricAggregation}.
+ * <p>
+ * {@link BaseMetric} is thread-safe and can be accessed by multiple threads.
+ */
+@Slf4j
+public class BaseMetric implements Metric {
+
+    public static String generateExpression(Class<? extends Aggregation> defaultAggregation) {
+        try {
+            Class<?> clazz = Class.forName(defaultAggregation.getCanonicalName());
+            Constructor<?> ctor = clazz.getConstructor();
+            Aggregation aggregation = (Aggregation) ctor.newInstance();
+            return aggregation.getAggFunctionFormat();
+        } catch (Exception exception) {
+            String message = String.format(
+                    "Cannot generate aggregation function for '%s'",
+                    defaultAggregation.getCanonicalName()
+            );
+            log.error(message, exception);
+            throw new IllegalStateException(message, exception);
+        }
+    }
+
+    private static final long serialVersionUID = 3055820948480283917L;
+
+    @Getter
+    private final String name;
+
+    @Getter
+    private final String longName;
+
+    @Getter
+    private final String description;
+
+    @Getter
+    private final Class<?> dataType;
+
+    @Getter
+    private final List<Class<? extends Aggregation>> aggregations;
+
+    @Getter
+    private final String metricExpression;
+
+    public BaseMetric(String metricField, Class<?> cls, EntityDictionary entityDictionary) {
+        Meta metaData = entityDictionary.getAttributeOrRelationAnnotation(cls, Meta.class, metricField);
+        Class<?> fieldType = entityDictionary.getType(cls, metricField);
+
+        this.name = metricField;
```

Comment:
I wonder if we need a base abstract class called Column that is common for both Dimension and Metric where we could consolidate the logic to parse the Meta annotation and set the basic fields?

Reply:
That's a great idea! I'd like to do it